pkg/kademlia/dht.go

```diff
@@ -26,16 +48,124 @@ func (k Kademlia) GetRoutingTable(ctx context.Context) (RoutingTable, error) {
 // Bootstrap contacts one of a set of pre defined trusted nodes on the network and
 // begins populating the local Kademlia node
 func (k Kademlia) Bootstrap(ctx context.Context) error {
-	return nil
+
+	dht, err := bkad.NewDHT(&bkad.MemoryStore{}, &bkad.Options{
+		BootstrapNodes: convertProtoNodes(k.bootstrapNodes),
+		IP:             k.ip,
+		Port:           k.port,
+		UseStun:        k.stun,
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := dht.CreateSocket(); err != nil {
+		return err
+	}
+
+	g := errgroup.Group{}
+
+	g.Go(func() error {
+		return dht.Listen()
+	})
+
+	g.Go(func() error {
+		return dht.Bootstrap()
+	})
+
+	return g.Wait()
 }
 
 // Ping checks that the provided node is still accessible on the network
-func (k Kademlia) Ping(ctx context.Context, node overlay.Node) (overlay.Node, error) {
-	return overlay.Node{}, nil
+func (k Kademlia) Ping(ctx context.Context, node proto.Node) (proto.Node, error) {
+	n, ok := convert(node).(*bkad.NetworkNode)
+	if !ok {
+		return proto.Node{}, NodeErr.New("unable to convert to expected type")
+	}
+	ok, err := k.dht.Ping(n)
+	if err != nil {
+		return proto.Node{}, err
+	}
+	if !ok {
+		return proto.Node{}, NodeErr.New("node unavailable")
+	}
+	return node, nil
 }
 
 // FindNode looks up the provided NodeID first in the local Node, and if it is not found
 // begins searching the network for the NodeID. Returns and error if node was not found
-func (k Kademlia) FindNode(ctx context.Context, ID NodeID) (overlay.Node, error) {
-	return overlay.Node{}, nil
+func (k Kademlia) FindNode(ctx context.Context, ID NodeID) (proto.Node, error) {
+	nodes, err := k.dht.FindNode([]byte(ID))
+	if err != nil {
+		return proto.Node{}, err
+
+	}
+
+	if len(nodes) <= 0 || string(nodes[0].ID) != string(ID) {
+		// check if the IDs don't match since dht.FindNode will
+		// return the closest node if the node it's looking for
+		// is not found
+		return proto.Node{}, NodeErr.New("node not found")
+	}
+
+	node := nodes[0]
+
+	return proto.Node{
+		Id: string(node.ID),
+		Address: &proto.NodeAddress{
+			Transport: defaultTransport, // TODO: defaulting to this, probably needs to be determined during lookup
+			Address:   fmt.Sprintf("%s:%d", node.IP.String(), node.Port),
+		},
+	}, nil
+}
+
+func convertProtoNodes(n []proto.Node) []*bkad.NetworkNode {
+	nn := []*bkad.NetworkNode{}
+	for i, v := range n {
+		if bnn, ok := convert(v).(*bkad.NetworkNode); !ok {
+			continue
+		} else {
+			nn[i] = bnn
+		}
+	}
+
+	return nn
+}
+
+func convertNetworkNodes(n []*bkad.NetworkNode) []proto.Node {
+	nn := []proto.Node{}
+	for i, v := range n {
+		if bnn, ok := convert(*v).(proto.Node); !ok {
+			continue
+		} else {
+			nn[i] = bnn
+		}
+	}
+
+	return nn
+}
+
+func convert(i interface{}) interface{} {
```

Comment:
since you usually immediately use this with an explicit typecast, instead of having the type switch and the type assert, maybe just have two convert functions? convertNode and convertNetworkNode? then you don't need a type switch or the type assert

Reply:
I had that originally but was duplicating the code and thought this looked cleaner but open to opinions

Reply:
oh? that's surprising to me. i don't think you'll duplicate any with what i'm saying. essentially if you have

```golang
func convertNetworkNode(i *bkad.NetworkNode) proto.Node {
		nn := proto.Node{
			Id:      string(v.ID),
			Address: &proto.NodeAddress{Transport: defaultTransport, Address: fmt.Sprintf("%s:%d", v.IP.String(), v.Port)},
		}
		return &nn
}
```

then instead of calling `convert(*v).(proto.Node)` where `*v` is already a `bkad.NetworkNode` you can just call `convertNetworkNode(v)` and no type assert or type switch at all

Reply:
(and vice versa for the other switch statement)

Reply:
ah ok I see what you are saying! I can make that change. Out of curiosity is their a downside to doing type switching ?

Reply:
ah ok I see what you are saying! I can make that change. Out of curiosity is their a downside to doing type switching ?

Reply:
no, the downside isn't type switching. type switching is great! especially if you're parsing an AST or something. the downside of the way the code is prior here is that convert takes an interface{}. with this change we get more compile-time type safety instead of runtime errors